clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -195,39 +196,55 @@ public Selector(long connectionMaxIdleMS, Metrics metrics, Time time, String met
      */
     @Override
     public void connect(String id, InetSocketAddress address, int sendBufferSize, int receiveBufferSize) throws IOException {
-        if (this.channels.containsKey(id))
-            throw new IllegalStateException("There is already a connection for id " + id);
-        if (this.closingChannels.containsKey(id))
-            throw new IllegalStateException("There is already a connection for id " + id + " that is still being closed");
+        connect(id, false, address, sendBufferSize, receiveBufferSize);
+    }
 
+    // Visible for testing only. Connecting in blocking mode allows us to verify the behavior when the connection
+    // is established by SocketChannel.connect
+    void blockingConnect(String id, InetSocketAddress address, int sendBufferSize, int receiveBufferSize) throws IOException {
```

Comment:
Would it be better to add a method `openSocketChannel(boolean shouldBlock)` or something along those lines that can be overridden in a test selector?

Reply:
Makes sense. Let me see if I can find a solution along these lines.